graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/TuplePrototypeBuiltins.java

```diff
@@ -0,0 +1,995 @@
+    }
+
+    public abstract static class JSTupleToStringNode extends JSBuiltinNode {
+
+        public JSTupleToStringNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected String toString(Tuple thisObj) {
+            return thisObj.toString();
+        }
+
+        @Specialization(guards = {"isJSTuple(thisObj)"})
+        protected String toString(DynamicObject thisObj) {
+            return JSTuple.valueOf(thisObj).toString();
+        }
+
+        @Fallback
+        protected void toStringNoTuple(@SuppressWarnings("unused") Object thisObj) {
+            throw Errors.createTypeError("Tuple.prototype.toString requires that 'this' be a Tuple");
+        }
+    }
+
+    public abstract static class JSTupleIteratorNode extends BasicTupleOperation {
+
+        @Child private ArrayPrototypeBuiltins.CreateArrayIteratorNode createArrayIteratorNode;
+
+        public JSTupleIteratorNode(JSContext context, JSBuiltin builtin, int iterationKind) {
+            super(context, builtin);
+            this.createArrayIteratorNode = ArrayPrototypeBuiltins.CreateArrayIteratorNode.create(context, iterationKind);
+        }
+
+        @Specialization
+        protected DynamicObject doObject(VirtualFrame frame, Object thisObj,
+                                         @Cached("createToObject(getContext())") JSToObjectNode toObjectNode) {
+            Tuple tuple = thisTupleValue(thisObj);
+            Object obj = toObjectNode.execute(tuple);
+            return createArrayIteratorNode.execute(frame, obj);
+        }
+    }
+
+    public abstract static class JSTupleValueOfNode extends JSBuiltinNode {
+
+        public JSTupleValueOfNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected Tuple doTuple(Tuple thisObj) {
+            return thisObj;
+        }
+
+        @Specialization(guards = "isJSTuple(thisObj)")
+        protected Tuple doJSTuple(DynamicObject thisObj) {
+            return JSTuple.valueOf(thisObj);
+        }
+
+        @Fallback
+        protected void fallback(@SuppressWarnings("unused") Object thisObj) {
+            throw Errors.createTypeError("Tuple.prototype.valueOf requires that 'this' be a Tuple");
+        }
+    }
+
+    public abstract static class JSTuplePoppedNode extends JSBuiltinNode {
+
+        public JSTuplePoppedNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected Tuple doTuple(Tuple thisObj) {
+            return getPoppedTuple(thisObj);
+        }
+
+        @Specialization(guards = "isJSTuple(thisObj)")
+        protected Tuple doJSTuple(DynamicObject thisObj) {
+            Tuple tuple = JSTuple.valueOf(thisObj);
+            return getPoppedTuple(tuple);
+        }
+
+        private Tuple getPoppedTuple(Tuple tuple) {
+            if (tuple.getArraySize() <= 1) {
```

Comment:
That `if` deserves a `ConditionProfile`. There is some chance that a certain `pop` call only ever sees empty tuples, and then you don't need to compile in all the code that follows (creating a tuple, copying array, ...).

Reply:
Added (and also in a very similar case, see `shifted`)